plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
 	}, nil
 }
 
+// selectNodesForPreemption finds all the nodes with possible victims for
+// preemption in parallel.
+func selectNodesForPreemption(pod *v1.Pod,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+	nodes []*v1.Node,
+	predicates map[string]algorithm.FitPredicate,
+	metadataProducer algorithm.MetadataProducer,
+) map[string][]*v1.Pod {
+
+	nodeNameToPods := map[string][]*v1.Pod{}
+	var resultLock sync.Mutex
+
+	// We can use the same metadata producer for all nodes.
+	meta := metadataProducer(pod, nodeNameToInfo)
+	checkNode := func(i int) {
+		nodeName := nodes[i].Name
+		pods, fits := selectVictimsOnNode(pod, meta, nodeNameToInfo[nodeName], predicates)
+		if fits {
+			resultLock.Lock()
+			nodeNameToPods[nodeName] = pods
+			resultLock.Unlock()
+		}
+	}
+	workqueue.Parallelize(16, len(nodes), checkNode)
+	return nodeNameToPods
+}
+
+// selectVictimsOnNode finds minimum set of pods on the given node that should
+// be preempted in order to make enough room for "pod" to be scheduled.
+// The algorithm first checks if the pod can be scheduled on the node when all the
+// lower priority pods are gone. If so, it sorts all the lower priority pods by
+// their priority and starting from the highest priority one, tries to keep as
+// many of them as possible while checking that the "pod" can still fit on the node.
```

Comment:
This algorithm works well if the infeasibility reason is insufficient resources or pod anti-affinity, but not if the reason is pod affinity. I think maybe for now it's best to just add a comment, as it only affects optimality, not correctness. The observations are:

(1) if the initial infeasibility reasons include pod affinity, then preemption will never help (removing existing pods will never make you satisfy pod affinity if you don't satisfy it initially). So you shouldn't even bother to try preemption when the infeasibility reason is pod affinity
and
(2) if the initial infeasibility reason is something else, say resources, then the short-circuit test ("If the new pod does not fit after removing all the lower priority pods, we are done...") can cause you to exit without trying any reprieving, when in fact reprieving some combination of pods might allow the pod to schedule (in particular, the pod(s) that the new pod requires to satisfy its affinity, as long as it/they can fit resource-wise along with the new pod).

If you wanted to address this, you could

(1) have this function exit immediately if any of the infeasibility reasons are pod affinity
and
(2) at the beginning of this function, check if the pod you are scheduling, or any of the existing pods on the node, have pod affinity, and if so, execute a different version of this function that tries all combinations of existing pods, or does something smarter (there are obviously smarter things you can do)

One other comment: Everything I said above is assuming node-level affinity and anti-affinity. For larger granularity, like rack or zone, I think the deal is
(i) For pod anti-affinity, I think the current algorithm works but is not optimal, in that it will miss an opportunity to preempt on (a) different node(s) than the one you are considering scheduling the pod onto, and thus may say the pod is infeasible everywhere in the cluster when actually it is not (or, will do more or higher-priority preemptions than if it preempted from other node(s)). This is just an optimality issue, so no need to deal with it now.
and
(ii) For pod affinity, I think the rack/zone/whatever case is the same as the single-node case, i.e. what I wrote above for pod affinity still applies.